mapie/multi_label_classification.py

```diff
@@ -0,0 +1,795 @@
+
+                    R = np.arange(self.n_lambdas) / self.n_lambdas
+                    K_R = np.cumsum(
+                        np.log(
+                            (
+                                1 -
+                                nu_batch *
+                                (losses_batch - R)
+                            ) +
+                            np.finfo(np.float64).eps
+                        ),
+                        axis=0
+                    )
+                    K_R = np.max(K_R, axis=0)
+                    K_R_max += K_R
+
+                r_hat_plus_tronc = self.lambdas[np.argwhere(
+                    np.cumsum(K_R_max > -np.log(delta), axis=1) == 1
+                )[:, 1]]
+                r_hat_plus = np.ones(self.n_lambdas)
+                r_hat_plus[:len(r_hat_plus_tronc)] = r_hat_plus_tronc
+
+        else:
+            r_hat_plus = (n_obs / (n_obs + 1)) * r_hat + (1 / (n_obs + 1))
+
+        return r_hat, r_hat_plus
+
+    def _find_lambda_star(
+        self,
+        r_hat_plus: NDArray,
+        alpha_np: NDArray
+    ) -> NDArray:
+        """Find the higher value of lambda such that for
+        all smaller lambda, the risk is smaller, for each value
+        of alpha.
+
+        Parameters
+        ----------
+        r_hat_plus : NDArray of shape (n_lambdas, )
+            Upper bounds computed in the get_r_hat_plus method.
+        alphas : NDArray of shape (n_alphas, )
+            Risk levels.
+
+        Returns
+        -------
+        NDArray of shape (n_alphas, )
+            Optimal lambdas which control the risks for each value
+            of alpha.
+        """
+
+        if len(alpha_np) > 1:
+            alphas_np = alpha_np[:, np.newaxis]
+        else:
+            alphas_np = alpha_np
+
+        bound_rep = np.repeat(
+            np.expand_dims(r_hat_plus, axis=0),
+            len(alphas_np),
+            axis=0
+        )
+        bound_rep[:, np.argmax(bound_rep, axis=1)] = np.maximum(
+            alphas_np,
+            bound_rep[:, np.argmax(bound_rep, axis=1)]
+        )
+        lambdas_star = np.argmin(
+                - np.greater_equal(
+                    bound_rep,
+                    alphas_np
+                ).astype(int),
+                axis=1
+            ) / len(r_hat_plus)
+
+        return lambdas_star
+
+    def partial_fit(
+        self,
+        X: ArrayLike,
+        y: ArrayLike,
+        _refit: Optional[bool] = False,
+        calib_size: Optional[float] = .3,
+    ) -> MapieMultiLabelClassifier:
+        """
+        Fit the base estimator or use the fitted base estimator on
+        batch data. All the computed risks will be concatenated each
+        time the partial_fit method is called.
+
+        Parameters
+        ----------
+        X : ArrayLike of shape (n_samples, n_features)
+            Training data.
+
+        y : NDArray of shape (n_samples, n_classes)
+            Training labels.
+
+        calib_size: Optional[float]
```

Comment:
Are you sure you want to allow estimator = None for partial_fit ?

Reply:
not allowed